teachertool/src/services/makecodeEditorService.ts

```diff
@@ -0,0 +1,52 @@
+interface PendingMessage {
+    original: pxt.editor.EditorMessageRequest;
+    handler: (response: any) => void;
+}
+
+let makecodeEditorRef: HTMLIFrameElement | undefined;
+const messageQueue: pxt.editor.EditorMessageRequest[] = [];
+let nextId: number = 0;
+let pendingMessages: {[index: string]: PendingMessage} = {};
+
+const onMessageReceived = (event: MessageEvent) => {
+    const data = event.data as pxt.editor.EditorMessageRequest;
+    if (data.type === "pxteditor" && data.id && pendingMessages[data.id]) {
+        const pending = pendingMessages[data.id];
+        pending.handler(data);
```

Comment:
Are there any messages the editor might send unbidden? As it is now, it seems `onMessageReceived` can only handle replies.

Reply:
If there are unbidden messages, are they things that we need to handle? Also, if there are messages that don't need give a response, is there anything else that we need to do? I don't know what those scenarios would look like.

Reply:
An example of an "unbidden" message would be one for when the workspace & content has loaded in the iframe, it'd be nice for it to send us a message saying so without us having to poll for it (I'm trying to do this now).

That said, I think this function would still receive those messages? We'd just need to make sure we handle the ones we're expecting even if they aren't in the pending messages queue. Since we don't currently have any we expect, I'm not sure if there's much to do here just yet. (Unless I'm missing something 🙂)

Reply:
Oh, thank you for that example, that makes sense. I do think that the unbidden case, then, should be handled when those cases arise. Or, I guess we could just have a notification dispatched saying that there was a message from the editor? Being that I'm not sure _what_ we'd want to do with those kinds of messages, I did nothing for now. The nice thing, though, is that it can be added easily. I am also unsure if there's much to do here currently, but I can add some default behavior like dispatching a notification or log something if it would be valuable for testing.

Reply:
Let's log unhandled messages to the console. See what surfaces.